### ISSUE TITLE
between resolves to the wrong method

### ISSUE THREAD

**Issue by jimtng** [boc-tothefuture/openhab-jruby#224]
See https://community.openhab.org/t/jruby-openhab-rules-system/110598/132

When `between` is called from inside an execution block (run/triggered/otherwise), it resolves to OpenHAB::DSL::Rules::RuleConfig::between (created by prop :between).

When `between` is called outside an execution block, it would resolve to TimeOfDay.between.

Compare:

```ruby
# This will say outside: OpenHAB::DSL::TimeOfDay::TimeOfDayRangeElement
logger.info "outside: #{between('10:00'..'18:00').begin.class}"

rule 'x' do
  on_start
  #this will say inside: String
  run { logger.info "inside: #{between('10:00'..'18:00').begin.class}" }
end
```

Out docs suggesting the use of between, which can be interpreted as being usable inside a rule execution block
https://boc-tothefuture.github.io/openhab-jruby/usage/misc/time_of_day/#between

I think the problem is our cucumber test were done outside a rule / execution block hence why it passed.

**Comment by pacive** [comment]
There was some namespacing confusion when the TimeOfDay module was renamed from TOD to TimeOfDay. Could this be the cause for the regression?

**Comment by boc-tothefuture** [comment]
They are stored as strings and converted on-demand because we don't want to pre-compute them in the rule because they get converted into two instants.  Inside of a rule and outside a rule config that would be OK. 

I am not actually sure why within the run context it is resolving to the rule_context.  

We store the actual caller of the rule block in the rule_config into a variable named @caller

We could try changing our run code from:
```
task.block.call(event)
```

to
```
@config.caller.instance_exec(event, task.block) do |event, block|
 block.call(event)
end
```
This should run the block in the context of the caller of the rule method.  
Obviously needs to be tested.  I am not sure if it would break any other assumptions.

**Comment by boc-tothefuture** [comment]
Has anyone started a branch to look at this?

**Comment by jimtng** [comment]
> Has anyone started a branch to look at this?

This is a bit beyond my current knowledge of Ruby

**Comment by boc-tothefuture** [comment]
Ok, I will start one.

**Comment by jimtng** [comment]
> ```
> @config.caller.instance_exec(event, task.block) do |event, block|
>  block.call(event)
> end
> ```
> 
> 
> This should run the block in the context of the caller of the rule method.
> 
> Obviously needs to be tested.  I am not sure if it would break any other assumptions.

First thing that came to mind is whether logger.info{"Rule #{name}") would still work inside the run block.

Another thing would be variable scopes... but that might be a good thing, so we can use instance variable as a "script-wide" variable?

**Comment by boc-tothefuture** [comment]
> First thing that came to mind is whether logger.info("Rule #{name}") would still work inside the run block.

It would not.

**Comment by boc-tothefuture** [comment]
resolved by #226
